pkg/apis/workflow/v1alpha1/data_types.go

```diff
@@ -0,0 +1,59 @@
+package v1alpha1
+
+// Data is a data template
+type Data struct {
+	PodPolicy PodPolicy `json:"podPolicy,omitempty" protobuf:"bytes,1,opt,name=podPolicy,casttype=PodPolicy"`
+
+	// Source sources external data into a data template
+	Source *DataSource `json:"source,omitempty" protobuf:"bytes,2,opt,name=source"`
+
+	// Transformation applies a set of transformations
+	Transformation Transformation `json:"transformation" protobuf:"bytes,3,rep,name=transformation"`
+}
+
+type PodPolicy string
+
+const (
+	PodPolicyAlways PodPolicy = "Always"
+)
+
+func (d *Data) UsePod() bool {
+	// If we're not using artifact paths, only use a pod if PodPolicy is set to Always
+	if d.Source == nil || d.Source.ArtifactPaths == nil {
+		return d.PodPolicy == PodPolicyAlways
+	}
+	return true
+}
+
+func (d *Data) GetArtifactIfAny() *Artifact {
+	if d.Source != nil && d.Source.ArtifactPaths != nil {
```

Comment:
I think you should not do this - couples types - I think the train-wreck pattern is fine here

Reply:
Done